Step 1: Load Dataset & Basic Overview

In [2]:
import numpy as np
import pandas as pd

In [3]:
sales_data = pd.read_csv(
    "../outputs/feature_engineered_sales_data.csv"
)

Check Dataset Shape

In [4]:
print("Dataset Shape:", sales_data.shape)

Dataset Shape: (1019516, 30)


Display First 5 Rows

In [5]:
sales_data.head()

,Order_ID,Product_ID,Product_Name,Customer_ID,Order_Date,Payment_Method,Order_Status,Quantity,Discount,Selling_Price_Details,...,Brand,Order_Value_Category,Discount_Category,Quantity_Category,Total_Orders,Purchase_Frequency_Category,Total_Spending,Spending_category,Customer_Segment,Profit_Category
0,ORD000001,PRD000116,Britannia Spinach 1L,CUST003379,2023-12-15,Credit Card,Delivered,3,0.00,175,...,Britannia,High Value,No Discount,Medium Quantity,188,High Frequency,171528.40,High Spender,Loyal High-Value customer,Medium Profit
1,ORD000001,PRD000344,Sensodyne Orange Juice 200g,CUST003379,2023-12-15,Credit Card,Delivered,1,0.00,50,...,Sensodyne,Low Value,No Discount,Low Quantity,188,High Frequency,171528.40,High Spender,Loyal High-Value customer,Low Profit
2,ORD000001,PRD000756,Dove Chana Dal Pack of 1,CUST003379,2023-12-15,Credit Card,Delivered,1,0.05,126,...,Dove,Low Value,Low Discount,Low Quantity,188,High Frequency,171528.40,High Spender,Loyal High-Value customer,Low Profit
3,ORD000002,PRD000878,Duracell Chocolate Cookies 1kg,CUST008843,2023-06-06,UPI,Delivered,1,0.15,40,...,Duracell,Low Value,Medium Discount,Low Quantity,40,High Frequency,33898.05,High Spender,Loyal High-Value customer,Low Profit
4,ORD000002,PRD000415,Kissan Potato Chips 200g,CUST008843,2023-06-06,UPI,Delivered,1,0.10,54,...,Kissan,Low Value,Low Discount,Low Quantity,40,High Frequency,33898.05,High Spender,Loyal High-Value customer,Low Profit


Check Column Names

In [6]:
sales_data.columns.to_list()

['Order_ID',
 'Product_ID',
 'Product_Name',
 'Customer_ID',
 'Order_Date',
 'Payment_Method',
 'Order_Status',
 'Quantity',
 'Discount',
 'Selling_Price_Details',
 'Cost_Price_Details',
 'Gross_Sales',
 'Discount_Amount',
 'Net_Sales',
 'Total_Cost',
 'Profit',
 'Profit_Margin',
 'Region',
 'Category_Details',
 'Sub_Category',
 'Brand',
 'Order_Value_Category',
 'Discount_Category',
 'Quantity_Category',
 'Total_Orders',
 'Purchase_Frequency_Category',
 'Total_Spending',
 'Spending_category',
 'Customer_Segment',
 'Profit_Category']

Check Data Types

In [7]:
sales_data.dtypes

Order_ID                           str
Product_ID                         str
Product_Name                       str
Customer_ID                        str
Order_Date                         str
Payment_Method                     str
Order_Status                       str
Quantity                         int64
Discount                       float64
Selling_Price_Details            int64
Cost_Price_Details               int64
Gross_Sales                      int64
Discount_Amount                float64
Net_Sales                      float64
Total_Cost                       int64
Profit                         float64
Profit_Margin                  float64
Region                             str
Category_Details                   str
Sub_Category                       str
Brand                              str
Order_Value_Category               str
Discount_Category                  str
Quantity_Category                  str
Total_Orders                     int64
Purchase_Frequency_Catego

Step 2: Overall Sales Performance

In [8]:
sales_data["Order_Date"] = pd.to_datetime(
    sales_data["Order_Date"],
    errors="coerce"
)

verify


In [9]:
print(sales_data["Order_Date"].dtypes)

datetime64[us]


Step 2.1: Calculate Overall KPIs

Total Sales

In [10]:
total_sales = sales_data["Net_Sales"].sum()
print("Total Sales:",round(total_sales,2))

Total Sales: 381805336.05


Total Orders

In [12]:
total_orders = sales_data["Order_ID"].nunique()
print("Total Orders:", total_orders)

Total Orders: 400000


Total Quantity Sold

In [13]:
total_quantity = sales_data["Quantity"].sum()
print("Total Quantity:",total_quantity)

Total Quantity: 1736819


Total Profit

In [14]:
total_profit = sales_data["Profit"].sum()
print("Total Profit:",total_profit)

Total Profit: 63932606.05


Average Order Value (AOV)

In [15]:
aov = total_sales/total_orders
print("Average Order Value:",round(aov, 2))

Average Order Value: 954.51


Step 2.2: Display All KPIs Together

In [18]:
kpi_summary = {
    "Total Sales": round(total_sales, 2),
    "Total Orders": total_orders,
    "Total Quantity": total_quantity,
    "Total Profit": round(total_profit, 2),
    "Average Order Value": round(aov, 2)
}

kpi_summary

{'Total Sales': np.float64(381805336.05),
 'Total Orders': 400000,
 'Total Quantity': np.int64(1736819),
 'Total Profit': np.float64(63932606.05),
 'Average Order Value': np.float64(954.51)}

Step 3: Sales Performance by Order Status

Total Sales by Status


In [26]:
sales_by_status = sales_data.groupby("Order_Status")["Net_Sales"].sum()
sales_by_status

Order_Status
Cancelled    1.733590e+07
Delivered    3.472383e+08
Failed       4.628679e+06
Pending      6.740408e+06
Returned     5.862087e+06
Name: Net_Sales, dtype: float64

Total Profit by Status

In [20]:
profit_by_status = sales_data.groupby("Order_Status")["Profit"].sum()
profit_by_status

Order_Status
Cancelled     2916854.75
Delivered    58117937.00
Failed         763319.45
Pending       1138937.75
Returned       995557.10
Name: Profit, dtype: float64

Total Quantity by Status

In [22]:
quantity_by_status = sales_data.groupby("Order_Status")["Quantity"].sum()
quantity_by_status

Order_Status
Cancelled      78520
Delivered    1580575
Failed         20707
Pending        30395
Returned       26622
Name: Quantity, dtype: int64

Total Orders by Status

In [23]:
orders_by_status = (
    sales_data.groupby("Order_Status")["Order_ID"]
    .nunique()
)

orders_by_status

Order_Status
Cancelled     18061
Delivered    363977
Failed         4739
Pending        7093
Returned       6130
Name: Order_ID, dtype: int64

Step 4: Sales Performance by Payment Method

Total Sales by Payment Method

In [27]:
sales_by_payment = (
    sales_data.groupby("Payment_Method")["Net_Sales"].sum()
)
sales_by_payment

Payment_Method
Cash           2.663096e+07
Credit Card    5.371062e+07
Debit Card     3.750557e+07
Net Banking    1.936567e+07
UPI            2.145044e+08
Wallet         3.008816e+07
Name: Net_Sales, dtype: float64

In [33]:
# Sort Sales in Descending Order
sales_by_payment.sort_values(ascending=False)

Payment_Method
UPI            2.145044e+08
Credit Card    5.371062e+07
Debit Card     3.750557e+07
Wallet         3.008816e+07
Cash           2.663096e+07
Net Banking    1.936567e+07
Name: Net_Sales, dtype: float64

Total Orders by Payment Method

In [30]:
orders_by_payment = (
    sales_data.groupby("Payment_Method")["Order_ID"]
    .nunique()
)

orders_by_payment

Payment_Method
Cash            28026
Credit Card     56129
Debit Card      39665
Net Banking     20113
UPI            224382
Wallet          31685
Name: Order_ID, dtype: int64

Step 5: Sales Performance by Region

Total Sales by Region

In [37]:
sales_by_region = (
    sales_data.groupby("Region")["Net_Sales"]
    .sum()
    .sort_values(ascending=False)
)
sales_by_region

Region
West       1.036566e+08
North      1.027134e+08
South      9.639006e+07
East       5.023940e+07
Central    2.880588e+07
Name: Net_Sales, dtype: float64

Total Profit by Region

In [38]:
profit_by_region  = (
    sales_data.groupby("Region")["Profit"]
    .sum()
    .sort_values(ascending=False)
)
profit_by_region

Region
West       17344441.20
North      17235767.65
South      16166000.90
East        8392412.75
Central     4793983.55
Name: Profit, dtype: float64

Format Sales Output

In [41]:
sales_by_region.apply(
    lambda x: f"{x:,.2f}"
)

Region
West       103,656,641.20
North      102,713,352.65
South       96,390,060.90
East        50,239,402.75
Central     28,805,878.55
Name: Net_Sales, dtype: str

Step 6: Sales Performance by Product Category

Total Sales by Category

In [42]:
sales_by_category = (
    sales_data.groupby("Category_Details")["Net_Sales"]
    .sum()
    .sort_values(ascending=False)
)
sales_by_category

Category_Details
Staples                67039618.80
Beverages              45099240.45
Snacks                 35896861.25
Fruits & Vegetables    35407781.40
Dairy & Breakfast      33003703.20
Packaged Food          28924217.70
Frozen Food            28333427.65
Bakery                 20664446.50
Home Care              19038626.40
Personal Care          18394545.20
Household Items        18216598.60
Beauty & Hygiene       17965096.70
Baby Care               8787096.70
Pet Care                5034075.50
Name: Net_Sales, dtype: float64

Total Profit By Category

In [45]:
profit_by_Category = (
    sales_data.groupby("Category_Details")["Profit"]
    .sum()
    .sort_values(ascending=False)
)
profit_by_region

Region
West       17344441.20
North      17235767.65
South      16166000.90
East        8392412.75
Central     4793983.55
Name: Profit, dtype: float64

Step 7: Product Performance Analysis

Total Sales by Product

In [46]:
sales_by_product = (
    sales_data.groupby("Product_Name")["Net_Sales"]
    .sum()
    .sort_values(ascending=False)
)
sales_by_product

Product_Name
Sensodyne Storage Bags Pack of 6       10740705.20
Nestle Kulfi Pack of 6                  8165271.00
Lay's Energy Drink 1L                   6808982.50
Mother Dairy Garam Masala Pack of 4     6508856.25
Parle Tea Pack of 4                     6287017.50
                                          ...     
Kurkure Mouthwash Pack of 6                1635.75
Everest Spinach 500g                       1620.00
Coca-Cola Aluminium Foil Pack of 2         1490.75
Sprite Masala Chips 500ml                  1482.00
Huggies Bhujia 200g                         438.00
Name: Net_Sales, Length: 1389, dtype: float64

Top 10 Products by Sales

In [47]:
top_10_products = sales_by_product.head(10)

top_10_products

Product_Name
Sensodyne Storage Bags Pack of 6       10740705.20
Nestle Kulfi Pack of 6                  8165271.00
Lay's Energy Drink 1L                   6808982.50
Mother Dairy Garam Masala Pack of 4     6508856.25
Parle Tea Pack of 4                     6287017.50
Lux Fruit Cake 500g                     5366515.00
Colgate Orange Drink 500ml              5228146.00
McCain Besan 1L                         4968734.00
Coca-Cola Low Fat Milk Pack of 4        4916973.60
Pedigree Turmeric Pack of 2             4449193.65
Name: Net_Sales, dtype: float64

Total Profit by Product

In [49]:
profit_by_product = (
    sales_data.groupby("Product_Name")["Profit"]
    .sum()
    .sort_values(ascending=False)
)
profit_by_product.head(10)

Product_Name
Colgate Orange Drink 500ml             1608926.00
Lay's Floor Cleaner Pack of 1          1067227.50
Parle Tea Pack of 4                    1030477.50
Gillette Bath Soap Pack of 2            868890.25
Mother Dairy Garam Masala Pack of 4     868646.25
Pedigree Turmeric Pack of 2             852433.65
Lux Fruit Cake 500g                     772915.00
Sensodyne Storage Bags Pack of 6        768185.20
McCain Besan 1L                         754394.00
Nestle Kulfi Pack of 6                  684496.00
Name: Profit, dtype: float64

Step 8: Category & Product Profitability Analysis

Profit Margin by Category

In [50]:
profit_margin_by_Category = (
    sales_data.groupby("Category_Details")["Profit_Margin"]
    .mean()
    .sort_values(ascending=False)
)
profit_margin_by_Category

Category_Details
Beauty & Hygiene       0.278030
Personal Care          0.261059
Snacks                 0.215035
Baby Care              0.203937
Home Care              0.202933
Dairy & Breakfast      0.194544
Bakery                 0.185650
Beverages              0.178167
Packaged Food          0.176129
Frozen Food            0.150254
Household Items        0.142273
Pet Care               0.140991
Fruits & Vegetables    0.116799
Staples                0.102640
Name: Profit_Margin, dtype: float64

Top 10 Products By Profit

In [53]:
top_10_profit_products = (
    sales_data.groupby("Product_Name")["Profit"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_10_profit_products

Product_Name
Colgate Orange Drink 500ml             1608926.00
Lay's Floor Cleaner Pack of 1          1067227.50
Parle Tea Pack of 4                    1030477.50
Gillette Bath Soap Pack of 2            868890.25
Mother Dairy Garam Masala Pack of 4     868646.25
Pedigree Turmeric Pack of 2             852433.65
Lux Fruit Cake 500g                     772915.00
Sensodyne Storage Bags Pack of 6        768185.20
McCain Besan 1L                         754394.00
Nestle Kulfi Pack of 6                  684496.00
Name: Profit, dtype: float64

Step 9: Customer Analysis

In [54]:
top_10_customers = (
    sales_data.groupby("Customer_ID")["Net_Sales"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)
top_10_customers

Customer_ID
CUST006820    2265194.80
CUST007952    1159569.20
CUST005370    1036419.45
CUST003340     952959.35
CUST005241     863664.90
CUST009220     811016.55
CUST003355     744092.15
CUST009017     731565.40
CUST001396     707867.00
CUST001110     697809.05
Name: Net_Sales, dtype: float64

Customer Order Frequency

In [56]:
customer_order = (
    sales_data.groupby("Customer_ID")["Order_ID"]
    .nunique()
    .sort_values(ascending=False)
)
customer_order.head(10)

Customer_ID
CUST006820    2305
CUST007952    1185
CUST005370    1082
CUST003340     994
CUST005241     872
CUST009220     861
CUST003355     773
CUST001396     761
CUST006743     748
CUST009017     746
Name: Order_ID, dtype: int64

Step 10: Monthly Sales Trends

In [60]:
sales_data["Month"] = sales_data["Order_Date"].dt.month

In [63]:
sales_data.columns

Index(['Order_ID', 'Product_ID', 'Product_Name', 'Customer_ID', 'Order_Date',
       'Payment_Method', 'Order_Status', 'Quantity', 'Discount',
       'Selling_Price_Details', 'Cost_Price_Details', 'Gross_Sales',
       'Discount_Amount', 'Net_Sales', 'Total_Cost', 'Profit', 'Profit_Margin',
       'Region', 'Category_Details', 'Sub_Category', 'Brand',
       'Order_Value_Category', 'Discount_Category', 'Quantity_Category',
       'Total_Orders', 'Purchase_Frequency_Category', 'Total_Spending',
       'Spending_category', 'Customer_Segment', 'Profit_Category', 'Month'],
      dtype='str')

Create Monthly Sales Analysis

Create Year-Month Column

In [65]:
sales_data["Order_Date"] = pd.to_datetime(
    sales_data["Order_Date"],
    errors="coerce"
)

sales_data["Year_Month"] = (
    sales_data["Order_Date"].dt.to_period("M")
)

Monthly Sales Analysis

In [66]:
monthly_sales = (
    sales_data.groupby("Year_Month")["Net_Sales"]
    .sum()
    .sort_index()
)

monthly_sales

Year_Month
2022-01     5371207.25
2022-02     4909445.05
2022-03     5546907.25
2022-04     5749336.15
2022-05     6288328.85
2022-06     5943573.70
2022-07     5719022.60
2022-08     5598357.20
2022-09     5930678.40
2022-10     6804875.80
2022-11     6437047.15
2022-12     6208490.80
2023-01     6288109.95
2023-02     5950866.40
2023-03     6894464.50
2023-04     7151974.40
2023-05     7603176.75
2023-06     7249619.05
2023-07     7067903.00
2023-08     7041083.95
2023-09     7043059.45
2023-10     8147862.10
2023-11     7773657.10
2023-12     7419624.25
2024-01     7676586.80
2024-02     7354546.80
2024-03     8355483.40
2024-04     8372966.55
2024-05     9115302.65
2024-06     8605287.20
2024-07     8557531.35
2024-08     8507613.30
2024-09     8741415.40
2024-10     9730669.00
2024-11     9242311.85
2024-12     8854336.15
2025-01     9016180.85
2025-02     8475266.10
2025-03    10051322.55
2025-04     9822315.30
2025-05    10673457.45
2025-06    10337091.85
2025-07    10149530.90


Monthly Sales DataFrame

In [67]:
monthly_sales_df = (
    sales_data.groupby("Year_Month")
    .agg(
        Total_Sales=("Net_Sales", "sum"),
        Total_Profit=("Profit", "sum"),
        Total_Orders=("Order_ID", "nunique")
    )
    .reset_index()
)

monthly_sales_df

,Year_Month,Total_Sales,Total_Profit,Total_Orders
0,2022-01,5371207.25,949872.25,5588
1,2022-02,4909445.05,861820.05,5161
2,2022-03,5546907.25,966217.25,5901
3,2022-04,5749336.15,928861.15,6104
4,2022-05,6288328.85,1022028.85,6631
5,2022-06,5943573.70,973153.70,6240
6,2022-07,5719022.60,995207.60,6054
7,2022-08,5598357.20,982362.20,5880
8,2022-09,5930678.40,965998.40,6254
9,2022-10,6804875.80,1110025.80,7140


Step 11: Quantity & Discount Analysis

Sales by Quantity Category

In [68]:
sales_by_quantity = (
    sales_data.groupby("Quantity_Category")["Net_Sales"]
    .sum()
    .sort_values(ascending=False)
)

sales_by_quantity

Quantity_Category
Low Quantity       2.624776e+08
Medium Quantity    1.193277e+08
Name: Net_Sales, dtype: float64

In [70]:
sales_by_quantity.apply(
    lambda x: f"{x:,.2f}"
)

Quantity_Category
Low Quantity       262,477,609.00
Medium Quantity    119,327,727.05
Name: Net_Sales, dtype: str

Sales by Discount Category

In [71]:
sales_by_discount = (
    sales_data.groupby("Discount_Category")["Net_Sales"]
    .sum()
    .sort_values(ascending=False)
)
sales_by_discount

Discount_Category
Low Discount       1.753910e+08
No Discount        1.450053e+08
Medium Discount    5.445308e+07
High Discount      6.955954e+06
Name: Net_Sales, dtype: float64

In [72]:
sales_by_discount.apply(
    lambda x: f"{x:,.2f}"
)

Discount_Category
Low Discount       175,391,024.35
No Discount        145,005,274.00
Medium Discount     54,453,083.45
High Discount        6,955,954.25
Name: Net_Sales, dtype: str

Profit by Discount Category

In [73]:
profit_by_discount = (
    sales_data.groupby("Discount_Category")["Profit"]
    .sum()
    .sort_values(ascending=False)
)

profit_by_discount

Discount_Category
No Discount        32099084.00
Low Discount       28593609.35
Medium Discount     3564683.45
High Discount       -324770.75
Name: Profit, dtype: float64

Step 12: Business Insights Summary

Regional Sales

In [74]:
region_summary = (
    sales_data.groupby("Region")
    .agg(
        Total_Sales=("Net_Sales", "sum"),
        Total_Profit=("Profit", "sum"),
        Total_Orders=("Order_ID", "nunique")
    )
    .reset_index()
)

region_summary

,Region,Total_Sales,Total_Profit,Total_Orders
0,Central,2.880588e+07,4793983.55,30141
1,East,5.023940e+07,8392412.75,52993
2,North,1.027134e+08,17235767.65,107697
3,South,9.639006e+07,16166000.90,101057
4,West,1.036566e+08,17344441.20,108112


Category Sales

In [75]:
category_summary = (
    sales_data.groupby("Category_Details")
    .agg(
        Total_Sales=("Net_Sales", "sum"),
        Total_Profit=("Profit", "sum"),
        Total_Quantity=("Quantity", "sum")
    )
    .reset_index()
    .sort_values("Total_Sales", ascending=False)
)

category_summary

,Category_Details,Total_Sales,Total_Profit,Total_Quantity
13,Staples,67039618.80,7592648.80,175220
3,Beverages,45099240.45,7670420.45,186434
12,Snacks,35896861.25,6185506.25,264947
6,Fruits & Vegetables,35407781.40,4113646.40,303203
4,Dairy & Breakfast,33003703.20,6484048.20,212514
9,Packaged Food,28924217.70,4837292.70,190964
5,Frozen Food,28333427.65,4171237.65,83416
1,Bakery,20664446.50,3891261.50,112961
7,Home Care,19038626.40,3710206.40,40915
10,Personal Care,18394545.20,4759510.20,49354


In [76]:
monthly_sales_df

,Year_Month,Total_Sales,Total_Profit,Total_Orders
0,2022-01,5371207.25,949872.25,5588
1,2022-02,4909445.05,861820.05,5161
2,2022-03,5546907.25,966217.25,5901
3,2022-04,5749336.15,928861.15,6104
4,2022-05,6288328.85,1022028.85,6631
5,2022-06,5943573.70,973153.70,6240
6,2022-07,5719022.60,995207.60,6054
7,2022-08,5598357.20,982362.20,5880
8,2022-09,5930678.40,965998.40,6254
9,2022-10,6804875.80,1110025.80,7140


Save Analysis Outputs

In [77]:
region_summary.to_csv(
    "../outputs/region_analysis.csv",
    index=False
)

category_summary.to_csv(
    "../outputs/category_analysis.csv",
    index=False
)

monthly_sales_df.to_csv(
    "../outputs/monthly_sales.csv",
    index=False
)

Step 13: Prepare Business Insights Summary

Top-Performing Region

In [78]:
top_region = region_summary.loc[
    region_summary["Total_Sales"].idxmax()
]

top_region

Region                 West
Total_Sales     103656641.2
Total_Profit     17344441.2
Total_Orders         108112
Name: 4, dtype: object

Find Top-Performing Category

In [79]:
top_category = category_summary.loc[
    category_summary["Total_Sales"].idxmax()
]

top_category

Category_Details       Staples
Total_Sales         67039618.8
Total_Profit         7592648.8
Total_Quantity          175220
Name: 13, dtype: object

Find Highest-Profit Category

In [80]:
highest_profit_category = category_summary.loc[
    category_summary["Total_Profit"].idxmax()
]

highest_profit_category

Category_Details      Beverages
Total_Sales         45099240.45
Total_Profit         7670420.45
Total_Quantity           186434
Name: 3, dtype: object

Find Highest-Sales Month

In [81]:
top_month = monthly_sales_df.loc[
    monthly_sales_df["Total_Sales"].idxmax()
]

top_month

Year_Month          2025-10
Total_Sales     11776907.45
Total_Profit     1944902.45
Total_Orders          12212
Name: 45, dtype: object

Business Insights

1. Regional Performance
- West region generated the highest total sales.
- West region recorded total sales of ₹103,656,641.20.
- The region generated total profit of ₹17,344,441.20.

2. Category Performance
- Staples was the highest sales-generating category.
- Staples generated total sales of ₹67,039,618.80.
- Beverages generated the highest total profit among the categories.
- Beverages generated total profit of ₹7,670,420.45.

3. Monthly Performance
- October 2025 recorded the highest total sales.
- Total sales in October 2025 were ₹11,776,907.45.
- The month recorded 12,212 unique orders.

4. Customer Analysis
- Customers were analyzed based on total spending and purchase frequency.
- Customer segments were created using spending and frequency categories.

5. Discount Analysis
- Sales and profit were compared across different discount categories.
- Discount performance can be further explored through visualizations.